crates/autopilot/src/infra/solvers/authenticator.rs

```diff
@@ -0,0 +1,107 @@
+    authenticator_eoa: ethcontract::Account,
+}
+
+pub struct Addresses {
+    pub settlement: Option<eth::Address>,
+    pub authenticator_eoa: eth::H256,
+}
+
+impl Authenticator {
+    /// Creates an authenticator which can remove solvers from the allow-list
+    pub async fn new(web3: DynWeb3, chain: ChainId, addresses: Addresses) -> Self {
+        let address_for = |contract: &ethcontract::Contract, address: Option<H160>| {
+            address
+                .or_else(|| deployment_address(contract, &chain))
+                .unwrap()
+        };
+
+        let settlement = contracts::GPv2Settlement::at(
+            &web3,
+            address_for(
+                contracts::GPv2Settlement::raw_contract(),
+                addresses.settlement.map(Into::into),
+            ),
+        );
+
+        let authenticator = contracts::GPv2AllowListAuthentication::at(
+            &web3,
+            settlement
+                .authenticator()
+                .call()
+                .await
+                .expect("authenticator address"),
+        );
+
+        let authenticator_role = contracts::Roles::at(
+            &web3,
+            deployment_address(contracts::Roles::raw_contract(), &chain).expect("roles address"),
+        );
+
+        Self {
+            authenticator,
+            authenticator_role,
+            authenticator_eoa: ethcontract::Account::Offline(
+                ethcontract::PrivateKey::from_raw(addresses.authenticator_eoa.0).unwrap(),
+                None,
+            ),
+        }
+    }
+
+    /// Fire and forget: Removes solver from the allow-list in the authenticator
+    /// contract. This solver will no longer be able to settle.
+    #[allow(dead_code)]
+    async fn remove_solver(&self, solver: domain::eth::Address) -> Result<(), Error> {
+        let calldata = self
+            .authenticator
+            .methods()
+            .remove_solver(solver.into())
+            .tx
+            .data
+            .ok_or(Error::SolverRemovalBadCalldata)?;
+        let authenticator_eoa = self.authenticator_eoa.clone();
+        let authenticator_address = self.authenticator.address();
+        let authenticator_role = self.authenticator_role.clone();
+        tokio::task::spawn(async move {
+            authenticator_role
+                .methods()
+                .exec_transaction_with_role(
+                    authenticator_address,
+                    0.into(),
+                    ethcontract::Bytes(calldata.0),
+                    0,
+                    ethcontract::Bytes([0; 32]), // @TODO: populate role
+                    true,
+                )
+                .from(authenticator_eoa)
+                .send()
+                .await
+                .map_err(Error::SolverRemovalFailed)
+        });
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("bad calldata for solver removal")]
+    SolverRemovalBadCalldata,
```

Comment:
I think it makes sense to limit the amount of error cases we export publicly. For example the case where we can't compute the correct calldata to send can't really happen (this just can't be expressed easily in the API of the library).
And the second error case can't be returned from this module because it's currently thrown in a future for which we never return the result for.